packages/protocol/contracts/L1/StakingProverPool3.sol

```diff
@@ -0,0 +1,316 @@
+        _;
+    }
+
+    modifier onlyProver(address prover) {
+        require(
+            stakers[proverToId[prover]].prover != address(0),
+            "Only provers can call this function"
+        );
+        _;
+    }
+
+    /**
+     * Initialize the rollup.
+     *
+     * @param _addressManager The AddressManager address.
+     */
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    // @Daniel's comment: Adjust the staking. Users can use this funciton to
+    // stake, re-stake, exit,
+    // and change parameters.
+    // @Dani: Most prob. not for exit and modify, because it would require extra
+    // logic and cases. Let handle
+    // modification in a separate function
+    function stake(
+        uint32 totalAmount,
+        address prover, // This can be a 'delegate' so not using msg.sender
+        uint16 rewardPerGas,
+        uint16 capacity
+    )
+        external
+        nonReentrant
+    {
+        // Cannot enter the pool below the minimum
+        require(minimumStake < totalAmount, "Cannot enter below minimum");
+        require(rewardPerGas > 0, "Cannot be less than 1");
+
+        // The list of 32 is not full yet so kind of everybody can enter into
+        // the pool
+        if (provers[31].stakedAmount == 0) {
+            stakers[currentIdIdx] = Staker(
+                currentIdIdx, prover, totalAmount, rewardPerGas, capacity
+            );
+            provers[currentIdIdx] = Prover(totalAmount, rewardPerGas);
+            proverToId[prover] = currentIdIdx;
+
+            // This would not overflow the 32 because of:
+            // provers[31].stakedAmount == 0
+            unchecked {
+                ++currentIdIdx;
+            }
+
+            // If we just filled up the last place in the array (so means full -
+            // we can
+            // determine which is the staker in risk)
+            if (provers[31].stakedAmount != 0) {
+                (minimumStakerId, minimumStake) = _determineLowestStaker();
+            }
+        } else {
+            // List is full, we need to check which position we need to put them
+            // into
+            (minimumStakerId,) = _determineLowestStaker();
+
+            // We need to overwrite it's place
+            stakers[minimumStakerId] = Staker(
+                minimumStakerId, prover, totalAmount, rewardPerGas, capacity
+            );
+            provers[minimumStakerId] = Prover(totalAmount, rewardPerGas);
+            proverToId[prover] = minimumStakerId;
+
+            // Need to determine again who is the last / lowest staker
+            (minimumStakerId, minimumStake) = _determineLowestStaker();
+        }
+
+        TaikoToken(AddressResolver(this).resolve("taiko_token", false)).burn(
+            prover, totalAmount
+        );
+
+        emit ProverEntered(prover, totalAmount, rewardPerGas, capacity);
+    }
+
+    function stakePositionAdjustments(
```

Comment:
anyone can use this function to adjust a stake position for another person, we need to keep track of a delagate on the Prover struct and set it to the original msg.sender, then check this on the `onlyProver` modifier?